explainers/signed-exchange-subresource-subtitution-explainer.md

```diff
@@ -0,0 +1,315 @@
+link HTTP headers (eg: Link: <https://cdn.publisher.example/lib.js>;
+rel="preload"; as="script"):
+   - Check whether matching "allowed-alt-sxg" link HTTP header
+   (`Link: <https://cdn.publisher.example/lib.js>;rel="allowed-alt-sxg";header-integrity="sha256-.."`)
+   exists or not. (Note that if the allowed-alt-sxg link HTTP header has
+   variants and variant-key attributes, the UA must execute the algorithm
+   written in
+   [HTTP Representation Variants spec](https://httpwg.org/http-extensions/draft-ietf-httpbis-variants.html)
+   to find the matching header.)
+   - Check whether all preload links which have matching allowed-alt-sxg link
+   header have matching (url and header-integrity) parsedExchange in
+   PrefetchedSignedExchangeCache. If the check passes, set request's stashed
+   exchange to the parsedExchange so the resource will be loaded from the cached
+   signed exchange.
+   - After processing the link HTTP headers, clears the
+   PrefetchedSignedExchangeCache.
+
+# Detailed design discussion
+## Header integrity of signed exchange
+We use the SHA256 hash value of *signedHeaders* in
+[application/signed-exchange format](https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#application-signed-exchange)
+for integrity checking. This signedHeaders is *"the canonical serialization of
+the CBOR representation of the response headers of the exchange represented by
+the application/signed-exchange resource, excluding the Signature header
+field"*. So this value doesn’t change even if the publisher signs the content
+again or changes the signing key. This header-integrity value can verify the
+integrity of both the response header and the content body, because signed
+exchange’s response header must have digest header.
+
+The UA needs to check that this value of the prefetched subresource signed
+exchange is same as the header-integrity attribute of allowed-alt-sxg link
+header in the response from the publisher. This is intended to prevent the
+distributor from encoding a tracking ID into the subresource signed exchange.
+
+We can’t use the
+[SRI’s integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity)
+for this purpose, because SRI’s integrity can be used only for verifying the
+integrity of the content body. So if the UA use the SRI’s integrity value in
+allowed-alt-sxg link header, we can use the subresource signed exchanges to
+track the users by changing content-type and detecting the image loading
+failure.
+
+## Multiple subresource signed exchanges
+If there are multiple matching subresource signed exchanges (example:
+script.js.sxg and image.jpg.sxg), the UA must check that there is no error in
+the all signed exchanges (eg: sig matching, URL matching, Merkle Integrity
+error) before processing the content of the signed exchanges. This is intended
+to prevent the distributor from encoding a tracking ID into the set of
```

Comment:
And to prevent the distributor from selecting a version of the subresource that isn't compatible with the selected version of the main resource, which might introduce a security hole.